### PR TITLE
HPCC-21493 Automate build tagging process

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
Used to tag 6.4.40 gold

### CODE REVIEW

**Comment by richardkchapman** [comment]
@ghalliday Please review

Targeting 6.4.x so that we can easily use it to tag any additional 6.4.x releases. Should not have any impact on code.

**Review by ghalliday — COMMENTED** [review]
This code doesn't quite work for new major/minor versions (unless we always start with a .2 release).  That can possibly be addressed later.

@richardkchapman a few minor points.

File: cmake_modules/parse_cmake.sh [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    ;;
+    -v|--verbose)
+    VERBOSE=-v
+    shift
+    ;;
+    -d|--dryrun)
+    DRYRUN=-d
+    shift
+    ;;
+    -r|--remote)
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply cmake_modules/parse_cmake.sh]: I will fix

File: cmake_modules/extract_tag.sh [hunk 2]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+# Generate the expected version tag(s) corresponding to the settings in version.cmake
+
+HPCC_HOME="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply cmake_modules/extract_tag.sh]: Changed to SCRIPT_DIR

File: cmake_modules/extract_tag.sh [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+. $HPCC_HOME/parse_cmake.sh
+
+for f in ${HPCC_PROJECT} ; do
+  set_tag
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply cmake_modules/extract_tag.sh]: Also missing the parse_cmake

Not sure whether this file is worth keeping, but I'll leave it for now.

File: cmake_modules/go_gold.sh [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+#
+
+HPCC_HOME="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+
+. $HPCC_HOME/parse_cmake.sh
+
+sync_git
+parse_cmake
+
+if [ "$HPCC_MATURITY" != "rc" ] ; then
```
[the remark this answers is not in the document]
Comment (richardkchapman) [reply cmake_modules/go_gold.sh]: Going gold on something that is already at release level implies a retag - which is unusual. I thin i originally allowed it if force was set. i will have a quick look.

[the remark this answers is not in the document]
Comment (richardkchapman) [reply cmake_modules/go_gold.sh]: Ok, added a --retag option

**Comment by richardkchapman** [comment]
I put in some code to handle the new minor release case - it's not perfect (and won't work for 7.2.x as we have not set the version info correctly - should really be called 7.1.0-closedown0 until we get to first rc). But it will be a trivial process to fix that manually when the time comes.

Major releases may need some manual intervention too, but I'm not so worried about that right now.

**Comment by richardkchapman** [comment]
@ghalliday

**Pull request merged**